### DISCUSSION TITLE
Find a way to make Hiccup forms easier to analyze

### DISCUSSION THREAD

**Discussion by Deraen** [reagent-project/reagent#620] [Ideas]
https://github.com/borkdude/clj-kondo/issues/25

- Optional way to create normal Reagent components from functions using some keyword, like `:n>` `:$`. Would work the same way as `[comp]`, but the keyword would make it possible to find the calls.
    - Namespaced keyword would be even better, but would be quite verbose.
- Provide function based API to optionally replace Hiccup syntax (like om.dom, helix.dom).

**Comment by borkdude** [comment]
Alternatives (not necessarily better than what's already mentioned):

- macro which expands into a vector call:

``` clojure
(defmacro f [& args] `[~@args])
(r/f component  1 2 3)
```

This solution doesn't need any changes to other code in Reagent, but allows static analysis to detect reagent calls accurately.

- clj-kondo could have a mode in which it always treats `[foo 1 2 3]` as a function call in .cljs files. False positives should be handled by users.

**Comment by Kaspazza** [comment]
Form-2 components are marked by clj-kondo as  unused binding 
```
(defn outer 
  [a b c]            ;; <--- those parameters will get marked by clj-kondo as unused-binding
  ;;  ....
  (fn [a b c]        ;; <--- they are repeated here
    [:div
      (str a b c)]))
```

**Comment by borkdude** [comment]
@Kaspazza The way to solve this is to write:

```
(defn outer [_a _b _c] ...)
```

**Comment by Kaspazza** [comment]
@borkdude Hmmm, I was in a mindset it is used so there should be no warning... But it makes sense... Thanks!
